### ISSUE TITLE
Undefined variable T using generic on Class (PEP695)

### ISSUE THREAD

**Issue by trentontrees** [pylint-dev/pylint#9335]
### Bug description

```python
# pylint: disable=missing-module-docstring
class Simple[T]:
  def __init__(self, value: T):
    self.value = value
```

### Command used

```shell
python3.12 -m pylint a.py
```

### Pylint output

```python
************* Module a
a.py:3:28: E0602: Undefined variable 'T' (undefined-variable)
```

### Expected behavior

There should be no error E0602 reported.

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.2
Python 3.12.1 (main, Dec  7 2023, 20:45:44) [Clang 15.0.0 (clang-1500.0.40.1)]
```

### OS / Environment

OSX - Sonoma 14.1.2 (23B92)

**Comment by jacobtylerwalls** [comment]
Thanks for the report.

Test case as a diff:
```diff
diff --git a/tests/functional/u/undefined/undefined_variable_py312.py b/tests/functional/u/undefined/undefined_variable_py312.py
index 0ca2475eb..88a542653 100644
--- a/tests/functional/u/undefined/undefined_variable_py312.py
+++ b/tests/functional/u/undefined/undefined_variable_py312.py
@@ -4,4 +4,5 @@ def f[T](a: T) -> T:
     print(a)
 
 class ChildClass[T, *Ts, **P]:
-    ...
+    def __init__(self, value: T):
+        self.value = value
```

**Comment by winstxnhdw** [comment]
I am getting the following error for generics on functions.

```python
Unused variable 'T'Pylint[W0612:unused-variable](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/unused-variable.html)
```

I am on pylint 3.1.0

**Comment by MildlyAggressiveFolk** [comment]
Same issue here.

### Pylint Version

```
pylint 3.1.0
astroid 3.1.0
Python 3.12.0  [Clang 13.0.0 (clang-1300.0.29.30)]
```


### OS / Environment
```
macOS Big Sur 11.7.8 (20G1351)
```

**Comment by lord-haffi** [comment]
Don't know if it's related to this one, but I'm getting the same error with
```python
class A[T]:
    pass

class B[T](A[T]):
    pass
```
It throws `E0602: Undefined variable 'T' (undefined-variable)` (only) in line `class B[T](A[T]):`.

```
pylint 3.2.6
astroid 3.2.4
Python 3.12.1
```

**Comment by Chris3606** [comment]
Same with the following class:
```py
class MyClass[T]:
    def func(self, x: T, y: T):
        print(x, y)
```

Version info:
```
pylint 3.2.6
astroid 3.2.4
Python 3.12.5 
```

**Comment by manuel-gallina** [comment]
Same issue with
```
pylint 3.3.1
astroid 3.3.5
Python 3.12.5
```
